render/render_objects.go

```diff
@@ -0,0 +1,234 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package render
+
+import (
+	"image"
+	"image/color"
+	"image/draw"
+	"math"
+
+	"github.com/lafriks/go-tiled"
+	"github.com/lafriks/go-tiled/internal/utils"
+
+	"github.com/disintegration/imaging"
+)
+
+// RenderVisibleGroups renders all visible groups
+func (r *Renderer) RenderVisibleGroups() error {
+	for _, group := range r.m.Groups {
+		if !group.Visible {
+			continue
+		}
+		if err := r._renderGroup(group); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RenderGroup renders single group.
+func (r *Renderer) RenderGroup(groupIdx int) error {
+	if groupIdx >= len(r.m.Groups) {
+		return ErrOutOfBounds
+	}
+
+	group := r.m.Groups[groupIdx]
+	return r._renderGroup(group)
+}
+
+func (r *Renderer) _renderGroup(group *tiled.Group) error {
+	for _, layer := range group.Layers {
+		if !layer.Visible {
+			continue
+		}
+		if err := r._renderLayer(layer); err != nil {
+			return err
+		}
+	}
+
+	for _, objectGroup := range group.ObjectGroups {
+		if !objectGroup.Visible {
+			continue
+		}
+		if err := r._renderObjectGroup(objectGroup); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RenderVisibleLayersAndObjectGroups render all layers and object groups, layer first, objectGroup second
+// so the order may be incorrect,
+// you may put them into different groups, then call RenderVisibleGroups
+func (r *Renderer) RenderVisibleLayersAndObjectGroups() error {
+	// TODO: The order maybe incorrect
+
+	if err := r.RenderVisibleLayers(); err != nil {
+		return err
+	}
+	return r.RenderVisibleObjectGroups()
+}
+
+// RenderVisibleObjectGroups renders all visible object groups
+func (r *Renderer) RenderVisibleObjectGroups() error {
+	for i, layer := range r.m.ObjectGroups {
+		if !layer.Visible {
+			continue
+		}
+		if err := r.RenderObjectGroup(i); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RenderObjectGroup renders a single object group
+func (r *Renderer) RenderObjectGroup(i int) error {
+	if i >= len(r.m.ObjectGroups) {
+		return ErrOutOfBounds
+	}
+
+	layer := r.m.ObjectGroups[i]
+	return r._renderObjectGroup(layer)
+}
+
+func (r *Renderer) _renderObjectGroup(objectGroup *tiled.ObjectGroup) error {
+	objs := objectGroup.Objects
+
+	// sort objects from left top to right down
+	objs = utils.SortAnySlice(objs, func(a, b *tiled.Object) bool {
+		if a.Y != b.Y {
+			return a.Y < b.Y
+		}
+
+		return a.X < b.X
+	})
+
+	for _, obj := range objs {
+		if err := r.renderOneObject(objectGroup, obj); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RenderGroupObjectGroup renders single object group in a certain group.
+func (r *Renderer) RenderGroupObjectGroup(groupIdx, objectGroupId int) error {
```

Comment:
As per lint suggestion:

```suggestion
func (r *Renderer) RenderGroupObjectGroup(groupIdx, objectGroupID int) error {
```